File: Leetcode/Medium/permutation-in-string.py
# https://leetcode.com/problems/permutation-in-string/
# 2020-06-07

class Solution(object):
        
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        
        if len(s1) > len(s2): return False
        if s1 == s2: return True
        
        def createMap(s, b, e):
            m = {}
            for i in range(b, e):
                c = s[i]
                m[c] = m[c] + 1 if c in m else 1
            return m
    
        # create hashmap to count letters in s1
        letters = createMap(s1, 0, len(s1))
        window = createMap(s2, 0, len(s1))
        if letters == window: return True 
        
        i = len(s1)
        while i < len(s2):            
            n = s2[i]
            if n not in letters:
                if i+1+len(s1) >= len(s2): return False
                window = createMap(s2, i+1, i+1+len(s1))
                i += len(s1)
            else:
                if n in window:
                    window[n] += 1
                else:
                    window[n] = 1

                r = s2[i - len(s1)]
                window[r] -= 1
                if window[r] == 0: del window[r]
                    
            if letters == window: return True
            i += 1
            
        return False

File: Leetcode/Medium/permutation-in-string.cpp
/*
https://leetcode.com/problems/permutation-in-string
2020-06-08
*/

class Solution {
public:
    bool checkInclusion(string s1, string s2) {

        if (s1 == s2) return true;
        if (s1.length() > s2.length()) return false;
            
        // use vector of numbers instead of map to optimize
        int letters[26] = {0};
        int offset = s1.length();     
            
        for (int i = 0; i < s1.length(); i++) {
            letters[s1[i]-'a']++;
        }
            
        // construct a sliding window
        int window[26] = {0};
        for (int i = 0; i < s2.size(); i++) {
            int curr = s2[i]-'a';
            window[curr]++;
            offset = (window[curr] <= letters[curr])? offset - 1 : offset + 1;
            
            if (i >= s1.length()) {
                int last = s2[i - s1.length()] - 'a';
                window[last]--;
                offset = (window[last] < letters[last]) ? offset + 1: offset - 1;
            }
            
            if (offset == 0) return true;
        }
        
        return false;
    }
};